### ISSUE TITLE
Discrepancy between yolov8 val results and pycocotools results (custom dataset). Which is the correct one?

### ISSUE THREAD

**Issue by ghost** [ultralytics/ultralytics#7120]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello. I know this is an issue that get posted around once a month here, but I can't seem to find a direct answer. I have trained a custom dataset on yolov8m and achieved some promising performance during validation. I am running this with the following standard script:

```
from ultralytics import YOLO

model = YOLO('path/to/weights/last.pt')
results = model.val(save_json=True)
```

I then realized that ultralytics doesn't provide AP_s and AP_m metrics, so I resorted to modifying the `detect/val.py` script so that pycocotools can read in my custom dataset in coco format. With the help of [this discussion](https://github.com/ultralytics/yolov5/issues/10747) and through some additional modifications, I was able to get some results:
![Screenshot from 2023-12-21 16-52-18](https://github.com/ultralytics/ultralytics/assets/42298486/25a42693-3716-4f49-901e-acda1cfe7cd0)
![Screenshot from 2023-12-21 16-52-35](https://github.com/ultralytics/ultralytics/assets/42298486/f6bb5880-c1cc-4119-aa09-e4c72f092a27)

As you can see, the mAP and mAP_50 results have a discrepancy of ~ 5%. After a bit of digging, I found [this conversation](https://github.com/ultralytics/yolov5/pull/7732) outlined the same problem but with yolov5. While this _did_  technically align the two sets of mAP metrics, it also made them both substantially worse.
![Screenshot from 2023-12-21 16-50-38](https://github.com/ultralytics/ultralytics/assets/42298486/63989009-d73f-41c1-b17a-972ee3f1835a)
![Screenshot from 2023-12-21 16-50-22](https://github.com/ultralytics/ultralytics/assets/42298486/7f5df3d2-21e4-4a7a-bb58-ff8c9dd01267)
I also presume there is a reason that this proposed alignment "fix" has not been pushed to yolov8.

I know a bit of variance is to be expected between the two - seeing that [this competition](https://github.com/ultralytics/yolov5/issues/2258) is still open. Even so, I am still confused as to why the difference is so big in the case of this custom dataset. When comparing the `predictions.json` file generated by `val.py` with `ground_truth.json`, the `category_ids` objects are both in the same format (1-10). Similarly, the `image_id` objects are also correct.

Yolo validation is run using the default values, meaning `max_det=300`. As far as I know, this is the only difference in hyperparameters between the two. And funnily enough, setting `max_det=100` to match pycocotools `maxDets=100` results in better performance.
![Screenshot from 2023-12-21 17-14-26](https://github.com/ultralytics/ultralytics/assets/42298486/05ffbd28-5489-4525-892e-e5d2585f63ff)

This leads me to the following four questions:

1.) Which set of metrics is more accurate? I am comparing these results with those from yolov7, and would like the most fair and accurate comparison. I know yolov7 is supported by a third-party, and they use pycocotools for evaluation, so I am leaning towards using the original pycocotools results.

2.) What are some reasons that could be causing this large discrepancy? I know it's been mentioned in other discussions that different techniques used could result in floating point differences, but I can't see how it would cause differences as large as over 4%. The only thing I can imagine happening is that the iou threshold parameter for nms in `cocoeval.py` of pycocotools is set much higher than the default for yolo (0.7). Unfortunately I can't seem to track that down in any of the pycocotool scripts.

3.) Why would setting `max_det` to a lower value lead to an increase in performance? Shouldn't the opposite take place? From my understanding, this variable controls the number of queries or predictions per image, which are then filtered out via nms. Shouldn't more predictions result in more potential, better guesses?

4.) Are there future plans of incorporating some of these more specialized mAP metrics into the yolo validation?

Also, I want to thank the ultralytics team for such great work. Your models are very intuitive and easy to work with.

**Comment by glenn-jocher** [comment]
@roncamposj hello! Thank you for your kind words and thorough investigation into this matter. To address your questions:

1.) The metrics that are most accurate would typically depend on how closely the metric evaluation aligns with your end application requirements. It's common to benchmark against pycocotools if you're looking to compare with other works that also use it as a standard. However, our validation results should theoretically be quite close, assuming identical evaluation settings.

2.) Large discrepancies can occur due to various reasons – preprocessing differences, evaluation script implementation (e.g., NMS thresholds, interpolation methods for AP calculation, etc.), or different handling of edge cases. The best way to approach this would be to ensure identical settings across tools for a fair comparison.

3.) Changing `max_det` might improve performance because it limits the number of detections fed into the NMS algorithm. This can sometimes weed out lower-quality predictions and keep only the top ones, resulting in a seemingly improved performance.

4.) We are always seeking to improve and meet community needs, and updating the metrics is a possibility. However, I can't give a specific timeframe for incorporating specialized mAP metrics. We recommend keeping up with our documentation and updates on our docs site which we've provided, where future enhancements will be detailed.

The focus should be on consistency across different tools, with awareness that small differences are natural due to the reasons mentioned above. Consistent results across different tools is an ongoing effort, and your feedback contributes to refining the process. Thank you for engaging with the community, and please continue to share your findings and suggestions! 🌟

**Comment by Johnny-zbb** [comment]
@glenn-jocher @roncamposj  Hi，I'm having the same problem, I'm comparing the segmentation of yolov8n-seg with other  segmentation models, I'm also using pycocotools . 
The accuracy of my yolov8-seg model trained based on a customized dataset is:
![image](https://github.com/ultralytics/ultralytics/assets/54133977/4cddf069-090c-4f4c-a099-9d0b79b1cfc9)
And I get pycocotools results based on the following parameters: conf=0.25,iou=0.6,max_det=100 as:
![image](https://github.com/ultralytics/ultralytics/assets/54133977/a8d382d9-6a17-4665-b370-5d42e419fb28)
and conf=0.25,iou=0.6,max_det=300：
![image](https://github.com/ultralytics/ultralytics/assets/54133977/4279fc8f-80a3-4053-8adc-23cd00873aa2)
Although I modified max_det, the precision didn't change.
I also tried conf=0.001,iou=0.6,max_det=100：
![image](https://github.com/ultralytics/ultralytics/assets/54133977/ce5e2a94-b6c7-4113-8c6d-c784c290362f)
While the recall has gotten higher, the precision has gotten lower, and is much worse than my yolo-seg precision.
And I've found that different iou and conf predictions get better coco accuracy ratings, but still don't reach the accuracy calculated by the ultralytics tool.

**Comment by Johnny-zbb** [comment]
@glenn-jocher Hello, above are the results I got with model.predict(). I experimented further and used the results of model.val(save_json=True) to evaluate the coco (segm) metrics which are already closer to the results evaluated by ultralytics, is the val function doing something special?

The results of model.val(save_json=True) were used to evaluate the coco metrics:
![image](https://github.com/ultralytics/ultralytics/assets/54133977/3276e2f4-5689-4c04-adcf-43c2bb239862)

I would also like to add that the coco segment format does not seem to affect the coco metrics when using the model.predict function. I started out using the cv2.findContours function to convert the mask to polygon format to store the predictions. But I realized that the val function of ultralytics saves the json using the RLE format. I thought findContours would result in losing some boundary information, so I standardized the format to RLE and still got poor evaluation results. Only when using the json data saved by the model.val() function, it comes close to the evaluation of ultralytics.

Thank you for your time and patience!

**Comment by glenn-jocher** [comment]
@Johnny-zbb hello! The `model.val()` function is designed to provide a standardized evaluation of your model's performance. It ensures consistent preprocessing, postprocessing (like NMS), and evaluation settings that are aligned with the training process. This consistency is crucial for fair and accurate performance assessment.

When you use `model.val(save_json=True)`, it saves the results in a format that is compatible with COCO evaluation tools, which might explain why the results are closer to what you expect. The difference in results when using `model.predict()` could be due to variations in how the data is processed or evaluated.

Regarding the segmentation format, both polygon and RLE are valid for COCO evaluation, but they may yield different results due to precision in boundary representation. However, if the discrepancy remains after standardizing to RLE, it suggests the difference may not be due to the format itself but potentially other factors in the evaluation pipeline.

It's always recommended to use the same evaluation script and settings that were used during model development for the most accurate comparison. If you continue to see significant discrepancies, please ensure that all settings (like confidence thresholds and IOU thresholds) are consistent across your evaluations.

Thank you for your detailed experimentation, and we appreciate your engagement with the YOLOv8 community! 🚀
